### DISCUSSION TITLE
Please stop overwriting releases

### DISCUSSION THREAD

**Discussion by mgorny** [aesara-devs/aesara#1453] [General]
I've noticed that the `rel-2.8.12` tag has been force-pushed for the third time. Please don't do that. It means that people aren't getting notified that there's a fix coming, and only learn when someone reports a checksum mismatch for the downloaded archive (which in turn mean users hit the problem head first). If the release is broken, then please make a new release with the fix.

**Comment by brandonwillard** [comment]
Do you mean in conda-forge or the tags here?

**Reply by mgorny** [reply]
Tags here.

**Reply by brandonwillard** [reply]
Weird, the last release should've been an entirely new tag.

**Reply by mgorny** [reply]
It was originally reported here: https://bugs.gentoo.org/895910

After comparing the file on Gentoo mirrors (at the time, I think it was overwritten by now) and fresh GitHub archive, there was a change in `.git_archival.txt`. Currently it has:

```
node: 404e2d4d45e98236aafb049e8b994d843957be41
node-date: 2023-02-21T12:56:43-06:00
describe-name: rel-2.8.12
ref-names: tag: rel-2.8.12
```

but originally one of the two bottom lines also had `HEAD -> ...` or something like that.

**Reply by mgorny** [reply]
I'm sorry, it just occurred to me that perhaps this is some kind of iffy git/GitHub behavior.

**Reply by brandonwillard** [reply]
Yeah, I definitely haven't been force pushing anything to `main` or the tags, so I assume this is coming from GitHub/CI.

**Reply by mgorny** [reply]
Ok, I am so sorry for blaming you in anger. Now I see that this is how git behaves on `$Format:%D$`. Apparently the archive originally contains `HEAD -> <branch>` when there are no new commits on top of the release tag, then that changes. I'm going to report this to setuptools_scm since the documented configuration causes this problem.

Once again, I am very sorry.

**Reply by brandonwillard** [reply]
No problem; if we are doing frustrating things, please do tell us!

**Reply by mgorny** [reply]
I've filed https://github.com/pypa/setuptools_scm/issues/806 for this, let's hope there's a good solution for it. Right now, the only "workaround" I can think of is making a commit immediately after the tag point (or making a release-specific commit for the tag and not pushing it to the main branch), so that tags never correspond to branch tips.

**Reply by brandonwillard** [reply]
CC: @maresb

**Reply by maresb** [reply]
Madness!!! Thanks for letting us know, and nice work tracking down the root cause.

Maybe we should include the tests with the sdist so that Gentoo could use the PyPI tarball?

**Reply by maresb** [reply]
CC: @carlsmedstad

**Reply by brandonwillard** [reply]
> Maybe we should include the tests with the sdist so that Gentoo could use the PyPI tarball?

Yeah, we can do that.

**Reply by mgorny** [reply]
> Maybe we should include the tests with the sdist so that Gentoo could use the PyPI tarball?

Yes, that would be great. Thank you! Perhaps also consider including documentation sources. While we don't use them at the moment, it is a recurring request from our users to install documentation that could be used offline.

**Reply by maresb** [reply]
Do you have a preferred mechanism for specifying the test environment? Namely, from where would you get the dependency list, and would you run the tests in a Conda env or virtualenv?

I'm slightly less enthusiastic about bundling the docs because it makes the sdist 5× bigger. But since size isn't a serious issue, bundling docs certainly isn't out of the question.

**Reply by maresb** [reply]
I just opened #1454 to add tests and docs to the sdist.

**Reply by mgorny** [reply]
> Do you have a preferred mechanism for specifying the test environment?

Not really. We "retype" dependencies anyway, so a format that's easy to diff would be best.

> Namely, from where would you get the dependency list, and would you run the tests in a Conda env or virtualenv?

Gentoo runs tests against installed system packages (since that's where aesara will be used when installed a system package). We run `python -m pytest` directly.
